src/lib/RoomConnection.ts

```diff
@@ -426,7 +445,11 @@ export default class RoomConnection extends TypedEventTarget {
         const remoteParticipant = new RemoteParticipant({ ...client, newJoiner: true });
         this.remoteParticipants = [...this.remoteParticipants, remoteParticipant];
         this._handleAcceptStreams([remoteParticipant]);
-        this.dispatchEvent(new CustomEvent("participant_joined", { detail: { remoteParticipant } }));
+        this.dispatchEvent(
```

Comment:
Should we move the state conversions up to the listeners in the `useRoomConnection` hook? If we ever want to reuse this `RoomConnection` class elsewhere we might want access to the full class, and it feels like a task that should be handled where we handle the rest of the react state

Reply:
I'm not sure if I understand correctly your point here. I only omit the props in the CustomEvent's payload, nothing changes in the `RoomConnection` class. 🤔 

Omitting the props here satisfies the TS better in this case, as it expects `RemoteParticipantState` in the payload, not `RemoteParticipant`.

```ts
var CustomEvent: new <{
    remoteParticipant: RemoteParticipantState;
}>(type: string, eventInitDict?: CustomEventInit<{
    remoteParticipant: RemoteParticipantState;
}> | undefined) => CustomEvent<...>
```

Reply:
So looking at the `RoomJoinedEvent` defined in `RoomConnection`, we see it's payload uses `RemoteParticipant`s
```typescript
// RoomConnection.ts line 71
export type RoomJoinedEvent = {
    localParticipant: LocalParticipant;
    remoteParticipants: RemoteParticipant[];
    waitingParticipants: WaitingParticipant[];
};
// line 138
export interface RoomEventsMap {
    ...rest;
    room_joined: (e: CustomEvent<RoomJoinedEvent>) => void;
}
// line 345
this.signalSocket.on("room_joined", this._handleRoomJoined.bind(this));
```

but then when we dispatch this `room_joined` event, we're converting to `RemoteParticipantState` prematurely, like so
```typescript
// RoomConnection.ts line 531
private _handleRoomJoined(event: SignalRoomJoinedEvent) {
    ...
    this.dispatchEvent(
        new CustomEvent("room_joined", {
            detail: {
                localParticipant: this.localParticipant,
                    remoteParticipants: this.remoteParticipants.map(
                        convertRemoteParticipantToRemoteParticipantState
                    ),
                    waitingParticipants: knockers.map((knocker) => {
                        return { id: knocker.clientId, displayName: knocker.displayName } as WaitingParticipant;
                    }),
                },
            })
        );
}
```
I think we need to move the state conversion up inside `useRoomConnection` like so:
```typescript
// useRoomConnection.ts line 475
createEventListener("room_joined", (e) => {
    const { localParticipant, remoteParticipants, waitingParticipants } = e.detail;
    dispatch({
        type: "ROOM_JOINED",
        payload: { 
            localParticipant, 
            remoteParticipants: remoteParticipants.map(convertRemoteParticipantToRemoteParticipantState), 
            waitingParticipants,
        },
    });
}),
```

hope that makes sense 😄

Reply:
you can see the error in typescript if you give the created custom event the `RoomJoinedEvent` type:
```typescript
            const customEvent: RoomJoinedEvent = new CustomEvent("room_joined", {
                detail: {
                    localParticipant: this.localParticipant,
                    remoteParticipants: this.remoteParticipants.map(convertRemoteParticipantToRemoteParticipantState),
                    waitingParticipants: knockers.map((knocker) => {
                        return { id: knocker.clientId, displayName: knocker.displayName } as WaitingParticipant;
                    }),
                },
            });
            this.dispatchEvent(customEvent);
```

...which probably means we need to type these events before dis to avoid this kind of error in the first place 😄

Reply:
Yes, this makes sense. Thanks a lot for the well detailed explanation! 🙏